### PR TITLE
Lint for unused borrows as part of UNUSED_MUST_USE

### LINKED ISSUE #76264: Warn when a borrow expression is unused
Borrowing an expression is side-effect free, so there should never be a reason to write the first instead of the second.

```rust
&expr;
// vs
expr;
```

If you really wanted to do this you could assign the result to an unused variable:

```rust
let _ = &expr;
```

Therefore, we should add a lint for unused borrows to the compiler. Why is this important? A misplaced semi-colon before an `&&` operator in a multi-line boolean expression causes subsequent clauses to be silently ignored.

```rust
let trex = TyrannosaurusRex::new();
let is_a_dog = has_a_tail(trex)
  && has_bad_breath(trex)
  && is_a_carnivore(trex); // Misplaced semi-colon (perhaps due to reordering of lines)
  && is_adorable(trex);

if is_a_dog {
    give_hug(trex); // Ouch!
}
```

### PR DESCRIPTION
**Pull request by 0xPoe**
close https://github.com/rust-lang/rust/issues/76264

base on https://github.com/rust-lang/rust/pull/76894

r? @RalfJung

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks for the PR!

I'm sorry, but I am pretty swamped currently and this is not in my core area of expertise.
@rust-lang/compiler could someone else take over reviewing this?

File: src/tools/clippy/tests/ui/iter_count.fixed [hunk 1]
```diff
@@ -3,11 +3,11 @@
 
 #![warn(clippy::iter_count)]
 #![allow(
-    unused_variables,
-    array_into_iter,
-    unused_mut,
-    clippy::into_iter_on_ref,
-    clippy::unnecessary_operation
+unused_variables,
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/iter_count.fixed:6]: This change and the same change in the `rs` file will fail Clippy tests. Please re-indent those.

Comment (hi-rustin) [reply src/tools/clippy/tests/ui/iter_count.fixed]: Fixed.

**Review by flip1995 — APPROVED** [review]
Clippy changes LGTM

**Comment by Aaron1011** [comment]
The FCP previously completed in https://github.com/rust-lang/rust/pull/76894#issuecomment-787090935.

Code LGTM.

@bors r+

**Pull request merged**